content/blog/2020-11-16-reproducibility-in-ml-my-tools/index.en.md

```diff
@@ -0,0 +1,126 @@
+
+I think the problem with all those solutions is twofold. 
+
+First, parameters need to be structured. That is, I want default settings that don't change much and other settings related to the same concepts grouped together. For example, my global training settings, such as my seed and my device (GPU), are two parameters that can be grouped into the same concept. YAML configuration file can solve this problem since one can structure information in such a way (see figure 1).
+
+``` yaml
+data_loader:
+    batch_size: 2048
+
+training_settings:
+    seed: 42
+    device: "cuda:0"
+```
+Figure 1:  Example of a YAML file.
+
+
+Second, parameters need to be hierarchical. I want to have specific parameters for a particular case without being forced to have them if they are not related to my case. For example, if a want to test my results using SGD and Adam optimizer, I will use two sets of parameters: a learning rate for SGD and a learning rate and betas value for Adam. If I'm using Argparse, I would need to have betas parameters even when using SGD; that just makes no sense to me. 
+
+[Hydra](https://hydra.cc/) solves both by hierarchically using YAML files. With Hydra, you can compose your configuration dynamically, enabling you to easily get the perfect configuration for each run. That is, you can enable settings for SGD when you use this one (`optimizer: SGD`) or Adam settings when using this one (`optimizer: Adam`) (see figure 2). This way of "calling" your configuration means that you will always get the settings you need and not those of other configurations. Also, this hierarchical structure is straightforward to understand, as shown in figure 3. We can see that we have 4 different models and can directly access them quickly without finding which one is used with this model and not this one since all of them only declare the parameters they need. For sure, if you only have 2 or 3 parameters, it seems overkill, but for how long will you only have 2 or 3 parameters? I don't know about you, but not much longer after I start a project, I rapidly have more than 3.
```

Comment:
perfect configuration for each run: pas clair. 
Ça semble dire que ça va te sortir les meilleurs hyperparamètres

Reply:
enable settings for SGD 
enable SGD settings

Reply:
this one: pas nécessaire dans les deux cas

Reply:
clarifier quand tu dis hierarchical structure. 

dire hierarchical parameter / configuration structure ?

pourraît être confus avec un modèle hierarchique.

Reply:
Figure

Reply:
We can see that we have 4 different
We can see four different

Reply:
quickly: redondant

Reply:
is used: passif

Reply:
which one is used with this model and not this one 
wordy, pas clair

Reply:
seems **like** overkill

Reply:
but for how long : pas clair.